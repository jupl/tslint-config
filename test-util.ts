import {readFile} from 'mz/fs'
import {
  Configuration,
  FormatterConstructor,
  IRuleFailureJson,
  Linter,
  RuleFailure,
} from 'tslint'

export interface ILintResult {
  readonly failures: ReadonlyArray<IRuleFailureJson>
  readonly errorCount: number
  readonly warningCount: number
  readonly fixes: ReadonlyArray<IRuleFailureJson>
  readonly format: string | FormatterConstructor
  readonly output: string
}

const basePath = process.cwd()
const configuration = Configuration.loadConfigurationFromPath('./tslint.json')
const linter = new Linter({fix: true}, Linter.createProgram('./tsconfig.json'))

export async function lintFixture(fixtureName: string): Promise<ILintResult> {
  const fileName = `./fixtures/${fixtureName}`
  const file = await readFile(fileName, 'utf8')
  linter.lint(fileName, file, configuration)
  const {failures, output, fixes = [], ...result} = linter.getResult()
  return {
    ...result,
    failures: failures.map(failureToJson),
    fixes: fixes.map(failureToJson),
    output: output.split(basePath).join('.'),
  }
}

function failureToJson(failure: RuleFailure): IRuleFailureJson {
  const {name, ...json} = failure.toJson()
  return {...json, name: name.split(basePath).join('.')}
}
