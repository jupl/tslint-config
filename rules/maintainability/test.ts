import {lintFixture} from '../../test-util'

test('Maintainability rules should be valid', async() => {
  const results = await lintFixture('maintainability.ts')
  expect(results).toMatchSnapshot()
})
