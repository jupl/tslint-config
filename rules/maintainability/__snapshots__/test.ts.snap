// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Maintainability rules should be valid 1`] = `
Object {
  "errorCount": 19,
  "failures": Array [
    Object {
      "endPosition": Object {
        "character": 14,
        "line": 56,
        "position": 825,
      },
      "failure": "There should be no space before \\"]\\"",
      "fix": Object {
        "innerLength": 1,
        "innerStart": 823,
        "innerText": "",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "array-bracket-spacing",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 13,
        "line": 56,
        "position": 824,
      },
    },
    Object {
      "endPosition": Object {
        "character": 5,
        "line": 49,
        "position": 714,
      },
      "failure": "Shadowed name: 'a'",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "no-shadowed-variable",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 4,
        "line": 49,
        "position": 713,
      },
    },
    Object {
      "endPosition": Object {
        "character": 10,
        "line": 57,
        "position": 837,
      },
      "failure": "Shadowed name: 'a'",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "no-shadowed-variable",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 9,
        "line": 57,
        "position": 836,
      },
    },
    Object {
      "endPosition": Object {
        "character": 49,
        "line": 12,
        "position": 254,
      },
      "failure": "ServerRequest is deprecated: Use IncomingMessage",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "deprecation",
      "ruleSeverity": "WARNING",
      "startPosition": Object {
        "character": 36,
        "line": 12,
        "position": 241,
      },
    },
    Object {
      "endPosition": Object {
        "character": 1,
        "line": 412,
        "position": 2660,
      },
      "failure": "file should end with a newline",
      "fix": Object {
        "innerLength": 0,
        "innerStart": 2660,
        "innerText": "
",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "eofline",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 1,
        "line": 412,
        "position": 2660,
      },
    },
    Object {
      "endPosition": Object {
        "character": 18,
        "line": 15,
        "position": 289,
      },
      "failure": "Expected linebreak to be 'LF'",
      "fix": Object {
        "innerLength": 1,
        "innerStart": 289,
        "innerText": "",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "linebreak-style",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 0,
        "line": 15,
        "position": 271,
      },
    },
    Object {
      "endPosition": Object {
        "character": 1,
        "line": 27,
        "position": 418,
      },
      "failure": "A maximum of 1 class per file is allowed.",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "max-classes-per-file",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 0,
        "line": 24,
        "position": 374,
      },
    },
    Object {
      "endPosition": Object {
        "character": 1,
        "line": 412,
        "position": 2660,
      },
      "failure": "This file has 413 lines, which exceeds the maximum of 400 lines allowed. Consider breaking this file up into smaller parts",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "max-file-line-count",
      "ruleSeverity": "WARNING",
      "startPosition": Object {
        "character": 0,
        "line": 412,
        "position": 2659,
      },
    },
    Object {
      "endPosition": Object {
        "character": 81,
        "line": 30,
        "position": 518,
      },
      "failure": "Exceeds maximum line length of 79",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "max-line-length",
      "ruleSeverity": "WARNING",
      "startPosition": Object {
        "character": 0,
        "line": 30,
        "position": 437,
      },
    },
    Object {
      "endPosition": Object {
        "character": 14,
        "line": 33,
        "position": 556,
      },
      "failure": "Use of default exports is forbidden",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "no-default-export",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 7,
        "line": 33,
        "position": 549,
      },
    },
    Object {
      "endPosition": Object {
        "character": 12,
        "line": 35,
        "position": 572,
      },
      "failure": "Multiple imports from 'net' can be combined into one.",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "no-duplicate-imports",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 0,
        "line": 35,
        "position": 560,
      },
    },
    Object {
      "endPosition": Object {
        "character": 5,
        "line": 38,
        "position": 601,
      },
      "failure": "Identifier 'g' is never reassigned; use 'const' instead of 'let'.",
      "fix": Object {
        "innerLength": 3,
        "innerStart": 596,
        "innerText": "const",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "prefer-const",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 4,
        "line": 38,
        "position": 600,
      },
    },
    Object {
      "endPosition": Object {
        "character": 5,
        "line": 43,
        "position": 671,
      },
      "failure": "Missing trailing comma",
      "fix": Object {
        "innerLength": 0,
        "innerStart": 671,
        "innerText": ",",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "trailing-comma",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 5,
        "line": 43,
        "position": 671,
      },
    },
    Object {
      "endPosition": Object {
        "character": 5,
        "line": 46,
        "position": 693,
      },
      "failure": "Missing trailing comma",
      "fix": Object {
        "innerLength": 0,
        "innerStart": 693,
        "innerText": ",",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "trailing-comma",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 5,
        "line": 46,
        "position": 693,
      },
    },
    Object {
      "endPosition": Object {
        "character": 13,
        "line": 49,
        "position": 722,
      },
      "failure": "Missing trailing comma",
      "fix": Object {
        "innerLength": 0,
        "innerStart": 722,
        "innerText": ",",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "trailing-comma",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 13,
        "line": 49,
        "position": 722,
      },
    },
    Object {
      "endPosition": Object {
        "character": 13,
        "line": 55,
        "position": 807,
      },
      "failure": "Unnecessary trailing comma",
      "fix": Object {
        "innerLength": 1,
        "innerStart": 806,
        "innerText": "",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "trailing-comma",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 12,
        "line": 55,
        "position": 806,
      },
    },
    Object {
      "endPosition": Object {
        "character": 12,
        "line": 56,
        "position": 823,
      },
      "failure": "Unnecessary trailing comma",
      "fix": Object {
        "innerLength": 1,
        "innerStart": 822,
        "innerText": "",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "trailing-comma",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 11,
        "line": 56,
        "position": 822,
      },
    },
    Object {
      "endPosition": Object {
        "character": 19,
        "line": 57,
        "position": 846,
      },
      "failure": "Unnecessary trailing comma",
      "fix": Object {
        "innerLength": 1,
        "innerStart": 845,
        "innerText": "",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "trailing-comma",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 18,
        "line": 57,
        "position": 845,
      },
    },
    Object {
      "endPosition": Object {
        "character": 20,
        "line": 57,
        "position": 847,
      },
      "failure": "missing whitespace",
      "fix": Object {
        "innerLength": 0,
        "innerStart": 846,
        "innerText": " ",
      },
      "name": "./fixtures/maintainability.ts",
      "ruleName": "whitespace",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 19,
        "line": 57,
        "position": 846,
      },
    },
    Object {
      "endPosition": Object {
        "character": 11,
        "line": 4,
        "position": 105,
      },
      "failure": "'a' is declared but never used.",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "no-unused-variable",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 10,
        "line": 4,
        "position": 104,
      },
    },
    Object {
      "endPosition": Object {
        "character": 11,
        "line": 5,
        "position": 120,
      },
      "failure": "'a' is declared but never used.",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "no-unused-variable",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 10,
        "line": 5,
        "position": 119,
      },
    },
    Object {
      "endPosition": Object {
        "character": 11,
        "line": 5,
        "position": 120,
      },
      "failure": "Mergeable namespace 'a' found. Merge its contents with the namespace on line 5.",
      "fix": undefined,
      "name": "./fixtures/maintainability.ts",
      "ruleName": "no-mergeable-namespace",
      "ruleSeverity": "ERROR",
      "startPosition": Object {
        "character": 10,
        "line": 5,
        "position": 119,
      },
    },
  ],
  "fixes": Array [],
  "format": "prose",
  "output": "
ERROR: ./fixtures/maintainability.ts[5, 11]: 'a' is declared but never used.
ERROR: ./fixtures/maintainability.ts[6, 11]: Mergeable namespace 'a' found. Merge its contents with the namespace on line 5.
ERROR: ./fixtures/maintainability.ts[6, 11]: 'a' is declared but never used.
WARNING: ./fixtures/maintainability.ts[13, 37]: ServerRequest is deprecated: Use IncomingMessage
ERROR: ./fixtures/maintainability.ts[16, 1]: Expected linebreak to be 'LF'
ERROR: ./fixtures/maintainability.ts[25, 1]: A maximum of 1 class per file is allowed.
WARNING: ./fixtures/maintainability.ts[31, 1]: Exceeds maximum line length of 79
ERROR: ./fixtures/maintainability.ts[34, 8]: Use of default exports is forbidden
ERROR: ./fixtures/maintainability.ts[36, 1]: Multiple imports from 'net' can be combined into one.
ERROR: ./fixtures/maintainability.ts[39, 5]: Identifier 'g' is never reassigned; use 'const' instead of 'let'.
ERROR: ./fixtures/maintainability.ts[44, 6]: Missing trailing comma
ERROR: ./fixtures/maintainability.ts[47, 6]: Missing trailing comma
ERROR: ./fixtures/maintainability.ts[50, 5]: Shadowed name: 'a'
ERROR: ./fixtures/maintainability.ts[50, 14]: Missing trailing comma
ERROR: ./fixtures/maintainability.ts[56, 13]: Unnecessary trailing comma
ERROR: ./fixtures/maintainability.ts[57, 12]: Unnecessary trailing comma
ERROR: ./fixtures/maintainability.ts[57, 14]: There should be no space before \\"]\\"
ERROR: ./fixtures/maintainability.ts[58, 10]: Shadowed name: 'a'
ERROR: ./fixtures/maintainability.ts[58, 19]: Unnecessary trailing comma
ERROR: ./fixtures/maintainability.ts[58, 20]: missing whitespace
WARNING: ./fixtures/maintainability.ts[413, 1]: This file has 413 lines, which exceeds the maximum of 400 lines allowed. Consider breaking this file up into smaller parts
ERROR: ./fixtures/maintainability.ts[413, 2]: file should end with a newline
",
  "warningCount": 3,
}
`;
